libs/optimizer/src/lib/optimizer.ts

```diff
@@ -0,0 +1,169 @@
+ * minTimeSec: beginning of the track (seconds elapsed since 01-01-1970T00:00:00.000)
+ */
+export interface ScoringTrack {
+  lat: number[];
+  lon: number[];
+  alt: number[];
+  timeSec: number[];
+  minTimeSec: number;
+}
+
+/**
+ * maxCycle: maximum duration in milliseconds for an optimization round trip. `
+ *           If undefined, calculation duration is unbounded.
+ * maxLoop: maximum number of iterations allowed for an optimization round trip.
+ *          If undefined, number of allowed iterations is unbounded
+ */
+export interface OptimizationOptions {
+  maxCycle?: number;
+  maxLoop?: number;
+}
+
+/**
+ * optimize function argument
+ * track: the ScoringTrack to optimize
+ * league: the LeagueCode of the leagues rules to follow. If undefined, defaults to LeagueCode.FFVL
+ * options: the OptimizationOptions for the computation
+ */
+export interface OptimizationRequest {
+  track: ScoringTrack;
+  league?: LeagueCode;
+  options?: OptimizationOptions;
+}
+
+// minimalistic for the moment. Will be improved in next iterations
+/**
+ * score: the score for the track in the given league
+ * optimal: the result is optimal (no need to get a next result of Iterator<OptimizationResult, OptimizationResult>)
+ */
+export interface OptimizationResult {
+  score: number;
+  optimal: boolean;
+}
+
+const ZERO_SCORE: OptimizationResult = {
+  score: 0,
+  optimal: true,
+};
+
+const MIN_POINTS = 5;
+
+/**
+ * computes the score for the flight
+ * @param request the OptimizationRequest
+ * @return an Iterator over the successive OptimizationResult
+ * @see README.md
+ */
+export function* optimize(request: OptimizationRequest): Iterator<OptimizationResult, OptimizationResult> {
+  if (request.track.lat.length == 0) {
+    console.warn('Empty track received in optimization request. Returns a 0 score');
+    return ZERO_SCORE;
+  }
+  const track = addPointsIfRequired(request.track);
+  const flight = toIgcFile(track);
+  const scoringRules = toScoringRules(request.league);
+  const options = toOptions(request.options);
+  const solutionIterator = solver(flight, scoringRules, options);
+  while (true) {
+    const solution = solutionIterator.next();
+    if (solution.done) {
+      console.debug(solution.value.processed, 'iterations');
+      return toResult(solution.value);
+    }
+    yield toResult(solution.value);
+  }
+}
+
+const NB_SEGMENTS_BETWEEN_POINTS = 2;
+
+/**
+ * the solver requires at least 5 points, so if there is not enough points,
+ * we create points between existing ones
+ */
+function addPointsIfRequired(track: ScoringTrack) {
+  if (track.lat.length >= MIN_POINTS) {
+    return track;
+  }
+  let newTrack: ScoringTrack = track;
+  while (newTrack.lat.length < MIN_POINTS) {
+    const segments: ScoringTrack[] = [];
+    for (let i = 1; i < newTrack.lat.length; i++) {
+      // split each segment of the track into two segments
+      segments.push(createSegments(getPoint(i - 1), getPoint(i), NB_SEGMENTS_BETWEEN_POINTS));
```

Comment:
I tried this solution, but it does not work.
To address the issue, I plan to return in the optimization result the effective track used for computation

Reply:
Another solution could be to compute intermediate points that are very close to the original points. In case we have to create these (few) points, retrieving the points in the original points should not be an issue: finding the nearest points in the original track will not be computationally expensive.

Reply:
Looking at https://github.com/mmomtchev/igc-xc-score/blob/99efcafcba4d4f8bbe11c8b95456bd8754e70ddb/src/flight.js#L121-L124

It looks like we should be able to repeat the last point if the timestamp is different

Reply:
I tried that but the results are not accurate and some tests fail. I have worked on a solution with intermediate points near the original points. That works.